### ISSUE TITLE
FP: `map_clone` suggests replacing an explicit `Foo::clone(x)` with `cloned()`

### ISSUE THREAD

**Issue by Zalathar** [rust-lang/rust-clippy#17550]
### Summary

The `clippy::map_clone` lint suggests replacing an explicit `.map(|x| Foo::clone(x)` with `.cloned()`, which is inappropriate for two reasons:
- If `x: &&Foo`, the call to clone is relying on auto-deref. Replacing the map-clone with `.cloned()` would call `<&Foo>::clone` instead of `<Foo>::clone`, changing the output type.
- The user explicitly wrote `Foo::clone(x)` instead of `x.clone()`, suggesting that cloning `Foo` specifically is important. Replacing the map-clone with `.cloned()` removes this layer of type-checking.
  - For example, `Arc::clone(x)` ensures that a cheap Arc clone is being performed.

### Lint Name

map_clone

### Reproducer

I tried this code:

```rust
pub fn foo(input: &[&String]) -> Vec<String> {
    input.iter().map(|s: &&String| String::clone(s)).collect()
}
```

I saw this happen:

```
warning: you are explicitly cloning with `.map()`
 --> src/lib.rs:2:5
  |
2 |     input.iter().map(|s: &&String| String::clone(s)).collect()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `input.iter().cloned()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.97.0/index.html#map_clone
  = note: `#[warn(clippy::map_clone)]` on by default
```

I expected to see this happen:

(no warning)

### Version

```text
rustc 1.99.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: aarch64-apple-darwin
release: 1.99.0-dev
LLVM version: 23.1.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Zalathar** [comment]
- https://github.com/rust-lang/rust-clippy/issues/10278 is somewhat similar, but doesn't include the explicit `Foo::clone(x)` call.

**Comment by Zalathar** [comment]
Encountered in the wild at https://github.com/rust-lang/rust/pull/160961#issuecomment-5262403476.

**Comment by Zalathar** [comment]
One thing to note is that the type signature on the closure argument shouldn’t make a difference either way; I only included it in the reproducer to make it easier to understand.
